### DISCUSSION TITLE
PhysX Integration via CMake

### DISCUSSION THREAD

**Discussion by mkomaritzan** [NVIDIA-Omniverse/PhysX#175] [Q&A]
I want to integrate physX into my own little geometry viewer application and tried to set this up with cmake. But unfortunately I found no guide for doing this for physX 5.x, only for 4.x versions, like this one: https://github.com/NVIDIAGameWorks/PhysX/issues/249

So I tried to adapt this guide above but my cmake configuration still fails...

I'm using Windows and my project structure is like main-project-dir/external/physx and in here ther is the stuff from the physx repo (e.g., bin, buildtools, compiler, ...). I've done the guide for windows and was also able to run the snippets so the physx setup itself should be fine, I hope...

In my CMakeLists.txt I added the following so far:

```
set(PHYSX_ROOT_DIR ${CMAKE_SOURCE_DIR}/external/physx)

set(TARGET_BUILD_PLATFORM "windows")
set(PLATFORM "Windows")

set(PX_BUILDSNIPPETS OFF CACHE BOOL "Generate the snippets")
set(PX_BUILDPUBLICSAMPLES OFF CACHE BOOL "Generate the samples projects")
set(PX_GENERATE_STATIC_LIBRARIES ON CACHE BOOL "Generate static libraries")
set(PX_FLOAT_POINT_PRECISE_MATH OFF CACHE BOOL "Float point precise math")
set(NV_USE_STATIC_WINCRT ON CACHE BOOL "Use the statically linked windows CRT")
set(NV_USE_DEBUG_WINCRT ON CACHE BOOL "Use the debug version of the CRT")
set(PX_OUTPUT_LIB_DIR ${CMAKE_LIBRARY_OUTPUT_DIRECTORY} CACHE INTERNAL "")
set(PX_OUTPUT_BIN_DIR ${CMAKE_RUNTIME_OUTPUT_DIRECTORY} CACHE INTERNAL "")


include_directories(${PHYSX_ROOT_DIR}/include)
add_subdirectory(${PHYSX_ROOT_DIR}/compiler/public external/physx)

#...

target_link_libraries(my_app other_libs ${PHYSX_LIBRARIES})
```

When I run this, I get some errors about dlls that cannot be copied like:
`
CMake Error at external/physx/source/compiler/cmake/windows/CMakeLists.txt:119 (FILE):
  FILE COPY cannot find "/bin/x86//PhysXDevice64.dll".
Call Stack (most recent call first):
  external/physx/source/compiler/cmake/CMakeLists.txt:103 (INCLUDE)
`

I know that the problem is, that it should actually look for this file at:
`/bin/win.x86_64.vc142.mt/<build-type>/PhysXDevice64.dll`

but I don't know what I need to set in the CMakeLists that it does use the correct path...

Any help would be very appreciated, I feel always a bit helpless in the 'setup' stage of a project :(

**Comment by preist-nvidia** [comment]
Hi @mkomaritzan - any chance the recently published vcpkg port could be helpful to you? https://omni-101.github.io/omniverse/2023/07/07/omniverse-physx-vcpkg-port/

**Reply by mkomaritzan** [reply]
I already tried that but that installation (`vcpkg.exe install physx`) also fails with
```physx is only supported on '(windows & x64 & !mingw & !uwp) | (linux & x64) | (linux & arm64)', which does not match x86-windows.```
I'm sure that I have a 64bit windows, don't know what the issue here is...

when I'm trying to run it with `--allow-unsupported` then it again failes with
```
CMake Error at ports/physx/portfile.cmake:60 (message):
  Unsupported platform/architecture combination
```

Thats what I meant with "I feel always a bit helpless in the 'setup' stage of a project". It always feels a bit jinxed when I'm trying to set something up...

**Reply by preist-nvidia** [reply]
I think you need to use `x64-windows`.
![image](https://github.com/NVIDIA-Omniverse/PhysX/assets/66367163/10a454de-b041-4900-a824-bb8fc909e041)

**Reply by mkomaritzan** [reply]
Wow, ok I should have read this more carefully... I thought this guide is only for what happens AFTER you installed the package via vcpkg...
using
`vcpkg.exe install physx --triplet x64-windows`

I was able to install physx and compile and run the small sample code. Now let's see if I can also combine it with my project code...
But that helped me alot already, thank you!!
